apps/passport-server/src/services/generic-issuance/pipelines/PODPipeline/PODPipeline.ts

```diff
@@ -0,0 +1,439 @@
+}
+
+export class PODPipeline implements BasePipeline {
+  public type = PipelineType.POD;
+  public capabilities: BasePipelineCapability[];
+
+  private eddsaPrivateKey: string;
+  private db: IPipelineAtomDB<PODAtom>;
+  private definition: PODPipelineDefinition;
+  private credentialSubservice: CredentialSubservice;
+  private consumerDB: IPipelineConsumerDB;
+  private cacheService: PersistentCacheService;
+
+  public get id(): string {
+    return this.definition.id;
+  }
+
+  public get feedCapability(): FeedIssuanceCapability {
+    return this.capabilities[0] as FeedIssuanceCapability;
+  }
+
+  public constructor(
+    eddsaPrivateKey: string,
+    definition: PODPipelineDefinition,
+    db: IPipelineAtomDB,
+    credentialSubservice: CredentialSubservice,
+    consumerDB: IPipelineConsumerDB,
+    cacheService: PersistentCacheService
+  ) {
+    this.eddsaPrivateKey = eddsaPrivateKey;
+    this.definition = definition;
+    this.db = db as IPipelineAtomDB<PODAtom>;
+    this.credentialSubservice = credentialSubservice;
+    this.consumerDB = consumerDB;
+    this.cacheService = cacheService;
+    this.capabilities = [
+      {
+        issue: this.feedIssue.bind(this),
+        options: this.definition.options.feedOptions,
+        type: PipelineCapability.FeedIssuance,
+        feedUrl: makeGenericIssuanceFeedUrl(
+          this.id,
+          this.definition.options.feedOptions.feedId
+        ),
+        getZuAuthConfig: async (): Promise<
+          PipelineEdDSATicketZuAuthConfig[]
+        > => []
+      } satisfies FeedIssuanceCapability
+    ] as unknown as BasePipelineCapability[];
+  }
+
+  /**
+   * Load the POD pipeline.
+   *
+   * Begins by loading input data from the definition, then creates atoms from
+   * the input and outputs. Finally saves those atoms and returns summary
+   * information.
+   *
+   * @returns A summary of the load operation
+   */
+  public async load(): Promise<PipelineLoadSummary> {
+    return traced(LOG_NAME, "load", async (span) => {
+      tracePipeline(this.definition);
+      logger(LOG_TAG, "load", this.definition.id, this.definition.type);
+
+      const start = new Date();
+      const logs: PipelineLog[] = [];
+
+      logger(
+        LOG_TAG,
+        `loading for pipeline id ${this.definition.id} with type ${this.definition.type}`
+      );
+      logs.push(
+        makePLogInfo(`loading data for pipeline '${this.definition.id}'`)
+      );
+
+      try {
+        const input = PODPipeline.loadInputFromDefinition(this.definition);
+
+        logs.push(
+          makePLogInfo(
+            `input columns: ${Object.keys(input.getColumns()).join(", ")}`
+          )
+        );
+        logs.push(makePLogInfo(`loaded ${input.getRows().length} rows`));
+
+        const atoms = PODPipeline.createAtoms(
+          input,
+          this.definition.options.outputs,
+          this.definition.id
+        );
+
+        // @todo because this code is async, it may end up being interleaved
```

Comment:
I'm definitely worried about this, particularly when combined with the DeleteFolder action below.  Also the presence of errors seems to leave a possibility of leaving the database empty when db.save() throws.

Up to whether you want to leave it for a later PR.  This kind of multi-part database update seems like it should be fixable with a transaction of some sort?  Depending on database+transaction settings, that likely deals with hiding intermediate states, and also rollback on errors.

Reply:
If it's a cache maybe these don't really need to be async?
Fine to defer solving a cross-cutting problem to outside of this PR, but I'm registering my concern if this doesn't get solved before we start relying on this pipeline in production.  I haven't checked if other pipelines have the same concern.

Reply:
This issue seems to overlap a lot with this one: https://github.com/proofcarryingdata/zupass/pull/1836, in that they are both concerned with the possibility that a user can request a feed at a moment when the "atom DB" is temporarily and unexpectedly empty.

The solution is likely to be the same, which is that we need to know if the atom DB is in a state where it's acceptable to read from it, possibly via the use of locks or a more explicit state machine.

Reply:
Having looked at this in more detail, I think we can solve it at the pipeline level by queueing all access of the DB, and making the `clear/save` operation atomic. I've pushed a new commit for this.